### ISSUE TITLE
Clang could warn about trailing spaces in `#include` filenames

### ISSUE THREAD

**Issue by zmodem** [llvm/llvm-project#96064]
Consider:

```
#include "foo.h " // Note: trailing space.
```

Assuming `foo.h` exists, the code above will compile on Windows whose filesystem API generally ignores leading and trailing space in filenames. However it will not compile on stricter systems such as Linux.

Maybe Clang should warn? It's niche, but also seems easy to do.

(Yes, we hit this in real code, https://chromium-review.googlesource.com/c/chromium/src/+/5641197)

**Comment by MitalAshok** [comment]
This also breaks `-Wnonportable-include-path`:

```c++
#include <array>
#include <array >
#include <Array >
#include <Array>
```

```
> clang-cl clang-cl test.cpp -fsyntax-only
test.cpp(4,10): warning: non-portable path to file '<array>'; specified path differs in case from file name on disk
      [-Wnonportable-include-path]
    4 | #include <Array>
      |          ^~~~~~~
      |          <array>
1 warning generated.
```

We should probably extend `-Wnonportable-include-path` to consider trailing spaces.

Dots at the end are ignored in the same way on Windows too:

```c++
// All of these include the same file
#include <Array >
#include <Array.>
#include <Array              >
#include <Array..............>
#include <Array.             >
#include <Array    . . ...  .>
```

Since these trailing spaces/dots are always removed when you specify a filename (except when prefixed by `\\?\`), we could issue the warning before even opening the file
https://learn.microsoft.com/en-gb/windows/win32/fileio/naming-a-file#naming-conventions (There's probably a more authoritative source on how the file name is transformed, this one only says "Do not end a file or directory name with a space or a period.")

**Comment by zmodem** [comment]
+1 extending `-Wnonportable-include-path` for this sounds good!

**Comment by taimoorzaeem** [comment]
@zmodem I would like to work on this. Could you assign it to me?

**Comment by akshaykumars614** [comment]
Let me understand the issue. 
You are saying
#include"foo.h " should throw a warning, but it is not?
I am I right? I am new here. Trying to understand things

**Comment by Jerrysun817** [comment]
I would like to work on this issue, can I get assigned please? @shafik

**Comment by charan-003** [comment]
hii, i want to work on this

**Comment by shivam-dhir** [comment]
Hi, I would love to work on this issue
